test-runner-poc/lib/runner.js

```diff
@@ -2,72 +2,122 @@
+
+  mongoClient.connect((err, client) => {
+    if (err) {
+      return environmentCallback(err);
+    }
+    createFilters(environmentParser);
+
+    function environmentParser(environmentName, version) {
+      console.log('environmentName ', environmentName);
+      const Environment = environments[environmentName];
+      parseConnectionString(mongodb_uri, (err, parsedURI) => {
+        if (err) {
+          return environmentCallback(err);
+        }
+        const environment = new Environment(parsedURI, version);
+        environment.mongo = require('../../index');
+        client.close(() => environmentCallback(err, environment));
+      });
+    }
+  });
 }
 
-before(function() {
+function createFilters(callback) {
+  let filtersInitialized = 0;
+  const filterFiles = fs.readdirSync(path.join(__dirname, 'filters'));
+
+  let topology, version;
+
+  filterFiles.filter(x => x.indexOf('js') !== -1).forEach(x => {
+    const FilterModule = require(path.join(__dirname, 'filters', x));
+    const filter = new FilterModule();
 
-	const environmentName = process.env['MONGODB_ENVIRONMENT'];
+    if (typeof filter.initializeFilter === 'function') {
+      filter.initializeFilter(_increment);
+    } else {
+      _increment();
+    }
 
-	console.log(`[environment: ${environmentName}]`);
+    //Makes sure to wait for all the filters to be initialized and added before calling the callback
+    function _increment() {
+      topology = topology || filter.runtimeTopology;
+      version = version || filter.mongoVersion;
+      filtersInitialized += 1;
+      addFilter(filter);
 
-	//apply filters
-	fs.readdirSync(path.join(__dirname, "filters"))
-		.filter(x => x.indexOf("js") !== -1)
-		.forEach(x => {
-			const FilterModule = require(path.join(__dirname, "filters", x));
-			addFilter(new FilterModule({ runtimeTopology: 'sharded' }));
-		});
+      if (filtersInitialized === filterFiles.length) {
+        callback(topology, version);
+      }
+    }
+  });
+}
 
+before(function(done) {
+  environmentSetup((err, environment) => {
+    if (err) {
+      done(err);
+    } else {
+      this.configuration = new TestConfiguration(environment);
+      done();
+    }
+  });
 });
 
-beforeEach(function() {
-	var self = this;
-
-	var called = 0;
-	function callback() {
-		called += 1;
-		if (called === filters.length) _run();
-	}
-
-	if (filters.length) {
-		filters.forEach(function(filter) {
-			callback();
-		});
-	}
-
-	function _run() {
-		if (!applyFilters(self.currentTest)) {
-			self.skip();
-		}
-	}
+beforeEach(function(done) {
+  // Assigned this to a new variable called self in order to preserve context and access tests within the _run function.
+  const self = this;
+  let filtersExecuted = 0;
+  if (filters.length) {
+    filters.forEach(function(filter) {
+      _run(filter);
+    });
+  }
+
+  function _run(filter) {
+    filtersExecuted += 1;
+
+    if (!filter.filter(self.currentTest)) {
+      if (!self.currentTest.parent.parent.root) {
```

Comment:
what's going on here? Can you leave a comment describing why we're checking this value?

Reply:
left a comment!